library/kani/src/aliasing.rs

```diff
@@ -0,0 +1,316 @@
+                }
+            }
+        }
+
+        /// Push a tag with a permission perm at pointer
+        pub(super) fn push<U>(tag: u8, perm: u8, pointer: *const U) {
+            // Decide whether to initialize the stacks
+            // for location:location+size_of(U).
+            // Offset has already been picked earlier.
+            unsafe {
+                use self::*;
+                if STATE == MonitorState::INIT
+                    && OBJECT == pointer_object(pointer)
+                    && OFFSET == pointer_offset(pointer)
+                {
+                    STACK_TAGS[STACK_TOP] = tag;
+                    STACK_PERMS[STACK_TOP] = perm;
+                    STACK_TOP += 1;
+                }
+            }
+        }
+
+        pub(super) fn stack_check<U>(tag: u8, access: bool, address: *const U) {
+            unsafe {
+                use self::*;
+                if STATE == MonitorState::INIT
+                    && OFFSET == pointer_offset(address)
+                    && OBJECT == pointer_object(address)
+                {
+                    let mut found = false;
+                    let mut j = 0;
+                    let mut new_top = 0;
+                    assert!(STACK_TOP < STACK_DEPTH);
+                    while j < STACK_DEPTH {
+                        if j < STACK_TOP {
+                            let id = STACK_TAGS[j];
+                            let kind = STACK_PERMS[j];
+                            if Permission::grants(access, kind) && id == tag {
+                                new_top = j + 1;
+                                found = true;
+                            }
+                        }
+                        j += 1;
+                    }
+                    STACK_TOP = new_top;
+                    STACK_VALID = STACK_VALID && found;
+                }
+            }
+        }
+    }
+
+    #[rustc_diagnostic_item = "KaniInitializeSState"]
+    pub fn initialize() {
+        self::monitors::prepare_monitors();
+    }
+
+    /// Push the permissions at the given location
+    pub fn push<U>(tag: u8, perm: u8, address: *const U) {
+        self::monitors::push(tag, perm, address)
+    }
+
+    /// Initialize the local stored at reference if initialized is set to false,
+    /// and track it using a monitor when using demonic nondeterminism.
+    ///
+    /// Every function call in the source program stack-allocates
+    /// the local variables that it uses; references to these
+    /// variables are only valid after these variables are initialized (first written).
+    /// Therefore this function can be used by supplying an initialized flag
+    /// set to true after the first write, a track flag set to the value
+    /// of a query to a demonic nondeterminism oracle (when this feature is used)
+    /// and a reference to the stack location.
+    #[rustc_diagnostic_item = "KaniInitializeLocal"]
+    pub fn initialize_local<U>(pointer: *const U) {
```

Comment:
I might be able to -- are you saying that we should do this to avoid instrumenting too many variables in the case that there are many variables but not many references or allocations? 

In the regression tests that I've written, this hasn't been a "rate-limiting factor" -- there are more references then there are "value" locals, and the cost of initializing a local is low compared to the cost of a stack check.